### DISCUSSION TITLE
How to package a zip file from a private artifactory that requires username and password

### DISCUSSION THREAD

**Discussion by tat2019** [microsoft/vcpkg#27113] [Q&A] [answered]
Hi folks,
I am new to vcpkg and I would like to ask a question on downloading a zipped third-party library to vcpkg from a private artifactory that requires authentication.

I have read this page, https://vcpkg.io/en/docs/examples/packaging-zipfiles.html, and I need to do something similar in linux environment. For example,

_vcpkg create some-lib http://my.articatory.com/some-lib-1.2.3.tar.gz some-lib-1.2.3.tar.gz_

Well, when I issued the above command, it didn't ask me for username and password. It seemed like the vcpkg create command was successful but under the vcpkg/downloads directory, the some-lib-1.2.3.tar.gz is not valid. Its size is not zero, and when I vi it, it shows something like the gz file can't be extracted.

To do something similar in the public domain like github, I created a private repo at github, with just one gz file in it.

_https://github.com/tat2019/tat_private_repo/raw/main/myfile.tar.gz_

So, I did something similar.

_#  ./vcpkg create myfile https://github.com/tat2019/tat_private_repo/raw/main/myfile.tar.gz myfile.tar.gz
-- Downloading https://github.com/tat2019/tat_private_repo/raw/main/myfile.tar.gz -> myfile.tar.gz...
CMake Error at scripts/ports.cmake:183 (message):
  Downloading
  https://github.com/tat2019/tat_private_repo/raw/main/myfile.tar.gz...
  Failed.  Status: 22;"HTTP response code said error"_

_# ls -al downloads/myfile.tar.gz
-rw-r--r-- 1 root root 0 Oct  5 23:35 downloads/myfile.tar.gz_

Why didn't it ask me for username and password?

Can anyone please tell me what I did wrong, or what I missed? I have spent a whole day googling but still couldn't find the answer.

Thanks in advance!

**Comment by Osyotr** [comment, marked as answer]
> Why didn't it ask me for username and password?

The whole point of automation is to remove possible interactive.

Check out `HEADERS` parameter for `vcpkg_download_distfile` https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_download_distfile.md#headers.

**Reply by tat2019** [reply]
> > Why didn't it ask me for username and password?
> 
> The whole point of automation is to remove possible interactive.
> 
> Check out `HEADERS` parameter for `vcpkg_download_distfile` https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_download_distfile.md#headers.

Thanks, Osyotr.

So, what should I put in the HEADERS field?

vcpkg_download_distfile(ARCHIVE
    URLS "https://someurl.com/artitfactory/myfile.tar.gz"
    FILENAME "myfile.tar.gz"
    SHA512 123...
    **HEADERS: MyUserName:MyPassword**
)

I tried but it didn't work when I tried to do "./vcpkg install myfile"

The download failed,

_curl: (22) The requested URL returned error: 401 Authorization Required_


Can you please give further instructions?

**Reply by Osyotr** [reply]
Pass curl header into `HEADERS` field, for example `"Authorization: Basic <Base64EncodedCredentials>"`, where `<Base64EncodedCredentials>` is the result of `echo $USERNAME:$PASSWORD | base64`.
There are other authorization methods that may be more suitable for you.
https://stackoverflow.com/q/3044315/5802715

**Reply by tat2019** [reply]
Thank you so much! It worked!
